src/txn/txn_timestamp.c

```diff
@@ -304,10 +304,10 @@ __wti_txn_update_pinned_timestamp(WT_SESSION_IMPL *session, bool force)
     __wti_txn_get_pinned_timestamp(
       session, &pinned_timestamp, WT_TXN_TS_ALREADY_LOCKED | WT_TXN_TS_INCLUDE_OLDEST);
 
-    if (pinned_timestamp != 0 &&
+    if (pinned_timestamp != WT_TS_NONE &&
       (!txn_global->has_pinned_timestamp || force ||
         txn_global->pinned_timestamp < pinned_timestamp)) {
-        txn_global->pinned_timestamp = pinned_timestamp;
+        WT_RELEASE_WRITE_WITH_BARRIER(txn_global->pinned_timestamp, pinned_timestamp);
         txn_global->has_pinned_timestamp = true;
```

Comment:
@ajmorton I think the has_pinned_timestamp write is also not correct. It is not defined as a shared variable, which is wrong. I don't want to fix it here. May be we can create a separate ticket.

Reply:
Yep that's likely. txn_global was scoped out of PM-3221 due to complexity and is scheduled for [SPM-3426](https://jira.mongodb.org/browse/SPM-3426). [WT-11149](https://jira.mongodb.org/browse/WT-11149) is probably the ticket for this.

Reply:
Can you move the `WT_RELEASE_WRITE_WITH_BARRIER` barrier onto `txn_global->has_pinned_timestamp`? 
`WT_RELEASE_WRITE_WITH_BARRIER` is defined as
```
#define WT_RELEASE_WRITE_WITH_BARRIER(v, val)   \
    do {                                        \
        WT_RELEASE_BARRIER();                   \
        __wt_atomic_store_generic(&(v), (val)); \
    } while (0)
```
so the current implementation is doing
```
RELEASE_BARRIER()
txn_global->pinned_timestamp = pinned_timestamp;
txn_global->has_pinned_timestamp = true
```
and not enforcing any ordering between the two writes. 
Similarly for the `WT_ACQUIRE_READ_WITH_BARRIER` call.

Reply:
I think we need both not only on the has_pinned_timestamp variable. But I don't want to increase the scope of this change so my initial idea is just to enforce we only read pinned_timestamp once.

Reply:
Agreed on `has_pinned_timestamp`. This is a clear case of message passing so I'd be happy to add the change in this PR. Perhaps we only add the ACQUIRE/RELEASE ops in `__wti_txn_update_pinned_timestamp` and `__wt_txn_pinned_timestamp` to keep the scope small? We can do the rest in the ticket to fix txn_global

Reply:
done